workshops/views.py

```diff
@@ -1672,48 +1683,65 @@ def instructors(request):
                 for lesson in lessons:
-                    instructors = instructors.filter(
+                    people = people.filter(
                         qualification__lesson=lesson
                     )
 
             if data['airport']:
                 x = data['airport'].latitude
                 y = data['airport'].longitude
                 # using Euclidean distance just because it's faster and easier
-                complex_F = ((F('airport__latitude') - x) ** 2
-                             + (F('airport__longitude') - y) ** 2)
-                instructors = instructors.annotate(distance=complex_F) \
-                                         .order_by('distance', 'family')
+                complex_F = ((F('airport__latitude') - x) ** 2 +
+                             (F('airport__longitude') - y) ** 2)
+                people = people.annotate(distance=complex_F) \
+                               .order_by('distance', 'family')
 
             if data['latitude'] and data['longitude']:
                 x = data['latitude']
                 y = data['longitude']
                 # using Euclidean distance just because it's faster and easier
-                complex_F = ((F('airport__latitude') - x) ** 2
-                             + (F('airport__longitude') - y) ** 2)
-                instructors = instructors.annotate(distance=complex_F) \
-                                         .order_by('distance', 'family')
+                complex_F = ((F('airport__latitude') - x) ** 2 +
+                             (F('airport__longitude') - y) ** 2)
+                people = people.annotate(distance=complex_F) \
+                               .order_by('distance', 'family')
 
             if data['country']:
-                instructors = instructors.filter(
+                people = people.filter(
                     airport__country__in=data['country']
                 ).order_by('family')
 
             if data['gender']:
-                instructors = instructors.filter(gender=data['gender'])
+                people = people.filter(gender=data['gender'])
 
             if data['instructor_badges']:
                 for badge in data['instructor_badges']:
-                    instructors = instructors.filter(badges__name=badge)
+                    people = people.filter(badges__name=badge)
+
+            # it's faster to count role=helper occurences than to check if user
+            # had a role=helper
+            if data['was_helper']:
+                people = people.filter(num_helper__gte=1)
+
+            if data['was_organizer']:
+                people = people.filter(num_organizer__gte=1)
+
+            if data['is_in_progress_trainee']:
+                q = Q(task__event__tags=TTT) & ~Q(task__event__tags=stalled)
+                people = (
+                    people.filter(q, task__role__name='learner')
+                    .exclude(badges__in=instructor_badges)
+                )
```

Comment:
Could you add a test for that in `test_searching_trainees`?
```python
        # Ironman
        Task.objects.create(person=self.ironman, event=e1, role=learner)
        Task.objects.create(person=self.ironman, event=e2, role=learner)

        ...

        self.assertEqual(list(response.context['persons']), [self.blackwidow, self.ironman])
```
This test fails (ironman is missing). However, if you execute the following code:
```python
        q = Q(task__event__tags=TTT) & ~Q(task__event__tags=stalled)
        print(Person.objects.filter(airport__isnull=False)
                            .select_related('airport')
                            .prefetch_related('badges', 'lessons')
                              # .annotate(
                              #       num_taught=Count(
                              #           Case(
                              #               When(task__role__name='instructor', then=Value(1)),
                              #               output_field=IntegerField()
                              #           )
                              #       ),
                              #       num_helper=Count(
                              #           Case(
                              #               When(task__role__name='helper', then=Value(1)),
                              #               output_field=IntegerField()
                              #           )
                              #       ),
                              #       num_organizer=Count(
                              #           Case(
                              #               When(task__role__name='organizer', then=Value(1)),
                              #               output_field=IntegerField()
                              #           )
                              #       )
                              #   )
                            .filter(q, task__role__name='learner')
                            .exclude(badges__in=Badge.objects.instructor_badges()))
```
you will get correct result (ironman is included). But when you uncomment the `annotate`, ironman is missing in the result. Looks like a bug in Django ORM?